### PR TITLE
No direct sim info for seed and track now

### PR DESCRIPTION
**Pull request by a-kapoor**
Got rid of all direct sim-info for creating seed and tracks. Instead fullsim global region and seed creator is used to create seeds. Moreover, track candidates are now initialized using back propagated seed state and not sim-track and sim-vertex information.

### CODE REVIEW

File: FastSimulation/Tracking/plugins/TrackCandidateProducer.cc [hunk 1]
```diff
@@ -85,7 +85,8 @@ TrackCandidateProducer::produce(edm::Event& e, const edm::EventSetup& es) {
 
   edm::ESHandle<Propagator>             propagator;
   es.get<TrackingComponentsRecord>().get(propagatorLabel,propagator);
-    
+  Propagator* thePropagator = propagator.product()->clone();
```
Comment (lveldere) [inline FastSimulation/Tracking/plugins/TrackCandidateProducer.cc]: don't clone
instead get a propagator that doesn't require change,
by using propagatorLabel "PropagatorWithMaterialOpposite"

File: FastSimulation/Tracking/plugins/TrackCandidateProducer.cc [hunk 2]
```diff
@@ -172,24 +173,19 @@ TrackCandidateProducer::produce(edm::Event& e, const edm::EventSetup& es) {
...
+
+    //Get seedTSOS from seed PTSOD//---------------------------------------------------------------------------
+    DetId seedDetId(seed.startingState().detId());
+    const GeomDet* gdet = trackerGeometry->idToDet(seedDetId);
+    TrajectoryStateOnSurface seedTSOS = trajectoryStateTransform::transientState(seed.startingState(), &(gdet->surface()),magneticField.product());
+    //---------------------------------------------------------------------------------------------------------
+    //backPropagate seedState to front recHit and get a new initial TSOS at front recHit//---------------------
     const GeomDet* initialLayer = trackerGeometry->idToDet(trackRecHits.front().geographicalId());
-    const TrajectoryStateOnSurface initialTSOS = propagator->propagate(initialFTS,initialLayer->surface()) ;
+    thePropagator->setPropagationDirection(oppositeToMomentum);
```
Comment (lveldere) [inline FastSimulation/Tracking/plugins/TrackCandidateProducer.cc]: get the right propagator from the start
(see earlier comment)

File: FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc [hunk 3]
```diff
@@ -115,35 +97,16 @@ TrajectorySeedProducer::TrajectorySeedProducer(const edm::ParameterSet& conf):
...
-        throw cms::Exception("FastSimulation/Tracking/TrajectorySeedProducer: bad configuration","Performance cut on SimTrack pT is tighter than cut on pT estimate from seed.");
-    }
-    if ((originHalfLength>=0 && simTrack_maxZ0>=0) && (originHalfLength>simTrack_maxZ0))
-    {
-        throw cms::Exception("FastSimulation/Tracking/TrajectorySeedProducer: bad configuration","Performance cut on SimTrack dz is tighter than cut on dz estimate from seed.");
-    }
-    if ((originRadius>=0 && simTrack_maxD0>=0) && (originRadius>simTrack_maxD0))
-    {
-        throw cms::Exception("FastSimulation/Tracking/TrajectorySeedProducer: bad configuration","Performance cut on SimTrack dxy is tighter than cut on dxy estimate from seed.");
+    if(conf.exists("RegionFactoryPSet")){
```
Comment (lveldere) [inline FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc]: don't do the 'if'
Assume RegionFactoryPSet is there.
If it's not there, cmsRun will crash, and that's how it's supposed to be.

File: FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc [hunk 4]
```diff
@@ -166,89 +129,17 @@ TrajectorySeedProducer::beginRun(edm::Run const&, const edm::EventSetup & es)
...
-    {
-        for (unsigned int iv = 0; iv < primaryVertices->size(); ++iv)
-        {
-            origins.push_back(&(*primaryVertices)[iv].position());
-        }
-    }
-
-    //only one possible origin is required to succeed
-    for (unsigned int i = 0; i < origins.size(); ++i)
+  if(theRegionProducer){
```
Comment (lveldere) [inline FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc]: don't do the 'if'
get rid of the else{*}

File: FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc [hunk 5]
```diff
@@ -166,89 +129,17 @@ TrajectorySeedProducer::beginRun(edm::Run const&, const edm::EventSetup & es)
...
-        if ((simTrack_maxD0>0.0) && ( theParticle.xyImpactParameter(origins[i]->X(),origins[i]->Y()) > simTrack_maxD0 ))
-        {
-            continue;
-        }
-        if ((simTrack_maxZ0>0.0) && ( fabs( theParticle.zImpactParameter(origins[i]->X(),origins[i]->Y()) - origins[i]->Z()) > simTrack_maxZ0))
-        {
-            continue;
-        }
-        return true;
+      std::cout<<"Either region producer, or the seed creator cfg is not available."<<std::endl;
```
Comment (lveldere) [inline FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc]: replace std::cout << "BLABLA" with
edm::LogWarning("TrajectorySeedProducer") << "BLABLA"

File: FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py [hunk 6]
```diff
@@ -1,5 +1,8 @@
 import FWCore.ParameterSet.Config as cms
 
+from RecoTracker.TkTrackingRegions.GlobalTrackingRegionFromBeamSpot_cfi import *
+from RecoTracker.TkSeedGenerator.SeedFromConsecutiveHitsCreator_cfi import *
+
 trajectorySeedProducer = cms.EDProducer("TrajectorySeedProducer",
                                         
     simTrackSelection = cms.PSet(
```
Comment (lveldere) [inline FastSimulation/Tracking/python/TrajectorySeedProducer_cfi.py]: get rid of the simTrackSelection here

File: FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc (hunk 5 above)
Comment (ANSH0712) [inline FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc]: But if the 'else' is no more there.. How would the warning be logged ?  The
code would just crash . Do you want it to crash or no ? If yes then I will
just remove the code and warning both.

Or do we want an exception there ?

On Fri, Aug 14, 2015 at 2:57 PM, lveldere <notifications@github.com> wrote:

> In FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc
> <https://github.com/cms-sw/cmssw/pull/10763#discussion_r37061680>:
>
> >      {
> > -        if ((simTrack_maxD0>0.0) && ( theParticle.xyImpactParameter(origins[i]->X(),origins[i]->Y()) > simTrack_maxD0 ))
> > -        {
> > -            continue;
> > -        }
> > -        if ((simTrack_maxZ0>0.0) && ( fabs( theParticle.zImpactParameter(origins[i]->X(),origins[i]->Y()) - origins[i]->Z()) > simTrack_maxZ0))
> > -        {
> > -            continue;
> > -        }
> > -        return true;
> > +      std::cout<<"Either region producer, or the seed creator cfg is not available."<<std::endl;
>
> replace std::cout << "BLABLA" with
> edm::LogWarning("TrajectorySeedProducer") << "BLABLA"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/10763/files#r37061680>.
>



-- 
ANSHUL KAPOOR
Research Student at IISER Pune
-----Please Don`t Spam----------
anshul.kapoor@cern.ch
anshul.kapoor@students.iiserpune.ac.in

Comment (lveldere) [inline FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc]: Hi Anshul

You are right, and I corrected a few sec after having made the comment.

In emails these updates get a mess,
so better check my comments on github.
Cheers

Lukas

Comment (Dr15Jones) [inline FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc]: The use of cout is not allowed in production code. Please change to using the message logger.

Comment (ANSH0712) [inline FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc]: Sure !!☺

On Fri, Aug 14, 2015 at 5:38 PM, lveldere <notifications@github.com> wrote:

> In FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc
> <https://github.com/cms-sw/cmssw/pull/10763#discussion_r37070297>:
>
> >      {
> > -        if ((simTrack_maxD0>0.0) && ( theParticle.xyImpactParameter(origins[i]->X(),origins[i]->Y()) > simTrack_maxD0 ))
> > -        {
> > -            continue;
> > -        }
> > -        if ((simTrack_maxZ0>0.0) && ( fabs( theParticle.zImpactParameter(origins[i]->X(),origins[i]->Y()) - origins[i]->Z()) > simTrack_maxZ0))
> > -        {
> > -            continue;
> > -        }
> > -        return true;
> > +      std::cout<<"Either region producer, or the seed creator cfg is not available."<<std::endl;
>
> Hi Anshul
>
> You are right, and I corrected a few sec after having made the comment.
>
> In emails these updates get a mess,
> so better check my comments on github.
> Cheers
>
> Lukas
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/10763/files#r37070297>.
>



-- 
ANSHUL KAPOOR
Research Student at IISER Pune
-----Please Don`t Spam----------
anshul.kapoor@cern.ch
anshul.kapoor@students.iiserpune.ac.in

**Comment by ANSH0712** [comment]
All issues solved ..

**Comment by lveldere** [comment]
@ANSH0712 please rebase

**Comment by ANSH0712** [comment]
Creating new pull request with all present conflicts resolved

**Pull request closed without merging**